### PR TITLE
Notion bug squash bounty

### PR DESCRIPTION
**Pull request by joshuatazrein**
Fixed all issues in the bounty as well as these feature requests:
- [Single line breaks](https://github.com/obsidianmd/obsidian-importer/issues/150)
- [Callouts](https://github.com/obsidianmd/obsidian-importer/issues/92)

### CODE REVIEW

File: src/main.ts [hunk 1]
```diff
@@ -9,6 +9,7 @@ import { NotionImporter } from './formats/notion';
 import { OneNoteImporter } from './formats/onenote';
 import { RoamJSONImporter } from './formats/roam-json';
 import { truncateText } from './util';
+import { NodePickedFile } from 'filesystem';
```
Comment (lishid) [inline src/main.ts:12]: Can we revert the changes to this file?

File: src/format-importer.ts [hunk 2]
```diff
@@ -147,7 +147,9 @@ export abstract class FormatImporter {
 	 * Recursively create folders, if they don't exist.
 	 */
 	async createFolders(path: string): Promise<TFolder> {
-		let normalizedPath = normalizePath(path);
+		// can't create folders starting with a dot
+		const sanitizedPath = path.replace(/^\./g, '').replace(/\/\./g, '/');
```
Comment (lishid) [inline src/format-importer.ts:151]:
```suggestion
		const sanitizedPath = path.split('/').map(n => n.replace(/^\.+/, '')).join('/');
```
Suggest changing to a split/join and processing each part separately, and also adding `+` to the regex in case there were multiple dots.

File: src/formats/notion.ts [hunk 3]
```diff
@@ -22,6 +25,13 @@ export class NotionImporter extends FormatImporter {
 			.addToggle((toggle) => toggle
 				.setValue(this.parentsInSubfolders)
 				.onChange((value) => (this.parentsInSubfolders = value)));
+
+		new Setting(this.modal.contentEl)
+		.setName('Single line breaks')
+		.setDesc('Separate Notion blocks with only one line break (default is 2).')
+		.addToggle((toggle) => toggle
+			.setValue(this.singleLineBreaks)
+			.onChange((value) => (this.singleLineBreaks = value)));
```
Comment (tgrosinger) [inline src/formats/notion.ts:34]:
```suggestion
			.onChange((value) => {
				this.singleLineBreaks = value;
			}));
```

File: src/formats/notion/convert-to-md.ts [hunk 4]
```diff
@@ -95,6 +108,10 @@ const typesMap: Record<NotionProperty['type'], NotionPropertyType[]> = {
 	],
 };
 
+function singleLineBreaks(markdownBody: string) {
```
Comment (tgrosinger) [inline src/formats/notion/convert-to-md.ts:111]: This probably doesn't need to be a separate function.

File: src/formats/notion/convert-to-md.ts [hunk 5]
```diff
@@ -217,13 +234,40 @@ function fixEquations(body: HTMLElement) {
 	}
 }
 
+function stripToSentence(paragraph: string) {
+	const firstSentence = paragraph.match(/^[^\.\?\!\n]*[\.\?\!]?/)?.[0]
+	return firstSentence ?? ''
+}
+
+function isCallout(element: Element) {
```
Comment (tgrosinger) [inline src/formats/notion/convert-to-md.ts:242]: Can this function be simplified to something like

```ts
return !!(/callout|bookmark/.test(element.getAttribute('class'))
```

I'm not seeing what the nullish is adding since both nullish values and an empty string are going to going to skip that clause.

File: src/main.ts (hunk 1 above)
Comment (joshuatazrein) [reply src/main.ts:12]: The current option for debugging that's commented out is outdated—I'll revert these changes but I'd recommend updating the commented-out debugging option for developers

File: src/formats/notion/convert-to-md.ts (hunk 5 above)
Comment (joshuatazrein) [reply src/formats/notion/convert-to-md.ts:242]: getAttribute('class') might return undefined, which breaks the regex.

**Comment by joshuatazrein** [comment]
I've resolved comments, let me know if there's anything else to address before merging.

**Comment by joshuatazrein** [comment]
Just checking if this is ready to merge!

**Review by tgrosinger — APPROVED** [review]

**Comment by tgrosinger** [comment]
Looks great, thank you.
